src/urllib3/multipart/decoder.py

```diff
@@ -0,0 +1,165 @@
+"""Logic for parsing and decomposing a multipart response body."""
+import email.parser
+import typing
+
+from .. import _collections
+from .encoder import encode_with
+
+if typing.TYPE_CHECKING:
+    from .. import response as _response
+
+
+@typing.overload
+def _split_on_find(content: str, bound: str) -> typing.Tuple[str, str]:
+    ...
+
+
+@typing.overload
+def _split_on_find(content: bytes, bound: bytes) -> typing.Tuple[bytes, bytes]:
+    ...
+
+
+def _split_on_find(
+    content: typing.AnyStr, bound: typing.AnyStr
+) -> typing.Tuple[typing.AnyStr, typing.AnyStr]:
+    point = content.find(bound)
+    return content[:point], content[point + len(bound) :]
+
+
+class ImproperBodyPartContentError(Exception):
+    pass
+
+
+class NonMultipartContentTypeError(Exception):
+    pass
+
+
+def _header_parser(
+    headers: bytes, encoding: str
+) -> typing.Sequence[typing.Tuple[str, str]]:
+    string = headers.decode(encoding)
```

Comment:
I'm not sure if we allow users to set the headers encoding elsewhere, will need to check.

Reply:
So to be very clear, this isn't Request headers, in `multipart/form-data` each part can have its own headers. And we're letting the user specify the encoding for handling the response bytes, so why wouldn't we use that for handling the headers too?

Reply:
Ah gotcha, that makes more sense since this `encoding` applies to the entire response body. Thanks for making this right in my head!